### PR TITLE
Solr harvester

### PR DESCRIPTION
**Pull request by bkiahstroud**
Ref https://github.com/harvard-lts/CURIOSity/issues/219

## Summary

Brings over Solr Harvesting logic from the [job_entry](https://github.com/harvard-lts/spotlight-oaipmh-resources/tree/job_entry) branch. 

These changes were heavily inspired from their original implementation, mostly found in this commit: e905c7c

However, I did need to make many changes in order to bring things up to date with Spotlight v3.3.0 and the updated `OaipmhHarvester`. 

<details><summary>Changes broken down by commit</summary>

## rename spotlight_oaipmh_harvesters to spotlight_harvesters

ca3f7b2b351b9451a98df3e9fc8ee8ae6bda8364

## add :type colum to harvesters table

034a40cd0717421e074d3f652aefe849ea72b1be

The :type column will allow us to use Single Table Inheritance for the OAIPMH and Solr harvesters.

The migration also includes a data migration to add a value for :type for all existing harvesters.

## create Harvester and SolrHarvester models

20a81770ea81e43284bb9c411067b57fc8bd892e

Sets up Harvester as the "base" class for the two harvest types. Migrates some common logic from OaipmhHarvester to Harvester inline with this.

## repurpose OaipmhHarvesterController to handle Solr harvests as well, add HarvestType

7e7423930ae1b083e0be9507ab94ea591be7754d

These changes are strongly inspired by changes found in the job_entry branch, originally authored by @ives1227

## update routes to use new "base" harvester model

0fb4fc60dc97c05f8425789feedc007395c33f43

## port SolrHarvester logic from job_entry branch (e905c7c)

fab3973d511e23cdd4134d6556f837ddaf934041

## rework HarvesterController for clarity

a3477871d32b75f9537a4556111bf45602904d3a

## add db column for solr harvest mapping filenames

bc22c77a21cc0ea12e6e45d24b41f0b7d2726586

## add solr option to harvester form

7811f11c9bf9af478c0f8546bd37d63270a59eb9

Builds UI that allows creation of SolrHarvesters. Combines changes from the deleted form partial with the one found in e905c7c

## locales from e905c7c

7009c99d6551372188d3c585c67ad0cdc62f4b99

Now that solr harvests are an option, this generalizes language on the harvester form to not just refer to MODS harvests

## extract common logic into base harvester model

412c905aeea5f91a2435703d09394ea177cf1bc6

## bring over default solr mapping file from job_entry branch

1da28fab2f7cbf944f1a71538e4e91eb724edfcf

## bring over unmodified logic relevant to solr harvests from the job_entry branch

bceed0098e7fc231e4b87b73c3b45808f697c89e

## rename SolrHarvestingItem to SolrHarvestingParser to better reflect its purpose

7b475c03ec71f2f4b270be14376e1fd465670fee

## move logic from SolrHarvestingBuilder into SolrHarvester

9c7418c28adc63480436f114f3096bb19806ccde

In Spotlight v3.3.0, the builder pattern is no longer used. This first step is extracting the existing logic (untouched so far) so we can delete the builder.

## create SolrUpload model

e017ee185f934b89b359ae5916d28d5853f5a369

This will be the SQL copy of items harvested by the SolrHarvester

## generalize #get_mapping_file method, remove most references to "resource" from SolrHarvester

f7d1d822c40e37f7a8def22840f3bbbd605cdf86

## persist harvested solr data in SolrUploads

940ed4cf5c07355e34ab4711e9d244b4dfc98345

## rename #get_harvests to #solr_harvests

1207f244dbdcd3c416251e9d5eb7d77e5d4e1816

## simplify getting data from solr

eb502820e9b20aefd3141564b370cdd89cfc1e1a

## properly connect to solr url

f56e1571297bb4529a83677e5b0b58b9edff2820

## align harvester method names

548d22a64602b04ba5d3ac49f49aeb31a1d46e91

## add job progress total logic to solr harvester

07ff377337673e0be4cf2b7ec1bc3cf67b0bc9d4

## finalize updating job tracker logic

35291cb5c0bcfa78eb6b9c18ced7a85bc8900f14

## remove unused cursor/schedule logic

1d7e65181411ed97d820c27ca7bec739efbb4406

Before, every "row" of solr data was harvested in a separate job. Due to how JobTrackers work in Spotlight v3.3.0 (one tracker per job), we run the whole solr harvest in a single PerformHarvestJob

## rearrange methods in SolrHarvester

600988674981a6ddb51f5fb608f89303f7273b89

## extract single item harvesting logic to SolrHarvester#harvest_item

0c6c2c7c26c9fca2f9f3e94837c562cc936dbf2a

## generalize common error handling logic

a9063113451f4fb536254761f1f7f8678b56fec6

## add @sidecar_id tracking to SolrHarvester

7ca2a2d90e77a9bc2006227067e569550414cf60

## fix namespace errors, simplify unique_id_field logic, remove unused #get_unique_id_field_name method

b60ac5a89de95f6e332712b7e5385ba5f8b6c355

## override SolrUpload#compound_id

00561a59d2c5be3bc8279d32756cd65b510dc295

This allows us to connect URNs with the correct SolrDocuments

## custom metadata gets indexed on initial solr harvest

7a98650437cadd15397af73733b8a3f8e1c6e2cf

## fix updating existing solr harvest items' metadata

71ec5a6d6e091b724d1da3d9f2d9a10bbd28b0ad

## account for missing trailing "/" in base_url

d65fb5e25bd9d70315d4695474a76ce29cac4595

</details>

## Local testing notes

Please `git checkout` [this CURIOSity branch](https://gitlab.com/notch8/harvard-curiosity/-/merge_requests/187) before bundling this version of the gem.

### CODE REVIEW

File: app/models/spotlight/solr_harvester.rb [hunk 1]
```diff
@@ -75,9 +75,10 @@ def harvest_item(record, job_tracker, job_progress)
       handle_item_harvest_error(e, parsed_solr_item, job_tracker)
     end
 
-    def solr_harvests(page = nil)
-      page = page.presence || 1
-      solr_connection.paginate(page, ROW_COUNT, 'select', params: { q: '*:*', wt: 'json' })
+    def solr_harvests(cursor = nil)
+      cursor = cursor.presence || '*'
+
+      solr_connection.get('select', params: { q: '*:*', cursorMark: cursor, sort: '_id asc', rows: ROW_COUNT, wt: 'json' })
```
Comment (bkiahstroud) [inline app/models/spotlight/solr_harvester.rb:81]: Is `_id asc` the correct way to sort the incoming results? `_id` seemed to be the only uniquely identifying field in the data I saw. 

cc @phil-plencner-hl @dl-maura @VanessaVenti

Comment (VanessaVenti) [reply app/models/spotlight/solr_harvester.rb:81]: That sounds right to me, unless Phil and Maura have other thoughts.

Comment (phil-plencner-hl) [reply app/models/spotlight/solr_harvester.rb:81]: ID seems like the best bet.

Comment (bkiahstroud) [reply app/models/spotlight/solr_harvester.rb]: See related commit: https://github.com/harvard-lts/spotlight-oaipmh-resources/pull/43/commits/9ea83ac1490c45f2d6ff5a62d08ab8a40cce3ceb

**Review by dl-maura — APPROVED** [review]
I was able to harvest teeming and acas-calendar

**Review by phil-plencner-hl — APPROVED** [review]
I was able to successfully harvest acats calendar using solr. Approved.

**Pull request merged**